### ISSUE TITLE
DESTROY is not called in a multithreaded environment.

### ISSUE THREAD

**Issue by vrurg** [rakudo/rakudo#2898]
## The Problem

Sometimes object's `DESTROY` submethod is not called if object is created in a multi-threaded environment.

## Steps to Reproduce

The following code demonstrates the bug:

```
my %objects;
my Lock $ol .= new;

class Bar {
    submethod TWEAK (|) {
        $ol.protect: {
            my $obj-id = self.WHICH;
            die "DUPLICATE FOUND: ", $obj-id if %objects{$obj-id};
            %objects{$obj-id} = True;
            say $obj-id;
        }
    }

    submethod DESTROY {
        $ol.protect: {
            my $obj-id = self.WHICH;
            say "DESTROY ", $obj-id;
            %objects{$obj-id}:delete;
        }
    }
}

my @p;
for 1..100 {
    for 1..100 {
        Bar.new;
        @p.push: start {
            # Bar.new;
            sleep .01;
         }
    }
}

await @p;
```

It doesn't matter wether a `Bar` object is created inside or outside of a thread. Running the code would eventually end up with something like:

> ```
> DUPLICATE FOUND: Bar|140310032644800
>   in submethod TWEAK at no-DESTROY.p6 line 10
>   in block  at no-DESTROY.p6 line 27
>   in block <unit> at no-DESTROY.p6 line 26
> ```

And with 100% assurance I can say that the object `Bar|140310032644800` would be found in the preceeding ouput but never paired with `DESTROY ` prefix.

## Environment
* Operating system: macOS and Linux
* Compiler version (`perl6 -v`): `This is Rakudo version 2019.03.1-391-ga12004f3e built on MoarVM version 2019.05-6-g61f0b7e80` but seemingly dates back as far as 2017.10 at least.

**Comment by vrurg** [comment]
So far, everything points out at moar because `DESTROY` is called by `finalize_handler` which in turn is not being called by GC.

**Comment by lizmat** [comment]
@vrurg: you know that there is **no** certainty that `DESTROY` will be called in an object, right?  Even with an orderly shutdown (which rakudo does **not** do by default, afaik), there is **no** guarantee that a `DESTROY` will be called.

**Comment by patzim** [comment]
I think that's not the problem vrurg is pointing at. IIUC the code correctly, the objects memory space is reused (indicating that the objects were dead), but `DESTROY` was never called on them.
I do think there is a guarantee that `DESTROY` is called before an object is freed and its memory reused (apart from shutdown, as you mentioned).

**Comment by jnthn** [comment]
>  IIUC the code correctly, the objects memory space is reused (indicating that the objects were dead), but DESTROY was never called on them.

Yes, that's my understand of the bug report, and I bet I can guess what it is, too. MoarVM runs the finalizer by sticking an exit handler onto the currently executing block. However, if it puts it on a VM thread that doesn't run userland code, or onto the thread pool's infinite loop, then those blocks never exit, so the finalizers are never triggered.

I wonder if we can somehow set up a finalizer thread and just post them all into a queue that it eats its way through.

**Comment by lizmat** [comment]
A finalizer thread seems like a good idea to me.

**Comment by dogbert17** [comment]
As far as I can tell this was fixed with 6b90262e8c36265
